woocommerce/woocommerce-admin
# Add tree shaking support for woocommerce components

Fixes #7024 

This adds tree shaking support to the `@woocommerce/components` package, allowing you to only bundle in the used components.
The biggest changes are the `sideEffects` addition in package.json, and the update in `extensions/build.js` to also build the style for each individual component if used outside of WooCommerce.

For the rest none of the components really looked like they had `sideEffects` , and I only updated a couple components that where in the same file.

No changelog as it is in the package.

### Detailed test instructions:

The easiest way of testing this is by making use of the extension examples, just have to update the webpack config for it a bit:
Update the `extensions/examples.config.js` by setting the mode to `production` (enabled tree shaking) adding components to the `excludedExternals` list for the extraction plugin:
```
new WooCommerceDependencyExtractionWebpackPlugin( {
	excludedExternals: [ 'components' ],
} )
```
And I also had to set `loose` to false for the babel-loader (I had some warnings show up), and disabled minification:
```
optimization: {
  minimize: false,
},
```
- Then build one of the examples `npm run example -- --ext=add-task` adding a `woocommerce/components` component to it somewhere. (The Spinner for example).
- Then check the `dist/index.js` file to see if only the imported component and it's dependencies are added, I used a grep command:
`grep -r 'CONCATENATED MODULE: ./packages/components' docs/examples/extensions/add-task/dist`
- Enable the plugin, and see if the component renders correctly (styling will work automatically given we are within WooCommerce).
- Now remove this line to remove the main styling: https://github.com/woocommerce/woocommerce-admin/blob/main/src/Loader.php#L420-L425
- Add the specific component styling to the example `index.scss`  (you might have to enqueue it as well, if it wasn't there before)
Going with the spinner example:
`@import '../../../../../packages/components/build-style/spinner/style.css';`
- Check if the component you added is styled correctly now.
